app/controllers/orders_controller.rb

```diff
@@ -238,71 +238,38 @@ def update
-        if update_order_details
-          quantities_after = @order.order_details.order("order_details.id").collect(&:quantity)
-
-          if quantities_after != quantities_before
-            flash[:notice] = "Quantities have changed, please review updated prices then click \"Purchase\""
-            return redirect_to order_path(@order)
-          end
-        else
-          logger.debug "errors #{@order.errors.full_messages}"
-          flash[:error] = @order.errors.full_messages.join("<br/>").html_safe
-          return render :show
-        end
-
-        # Empty message because validate_order! and purchase! don't give us useful messages as to why they failed
-        raise NUCore::PurchaseException.new("") unless @order.validate_order! && @order.purchase!
-
-        if facility_ability.can? :order_in_past, @order
-          raise NUCore::PurchaseException.new(I18n.t("controllers.orders.purchase.future_dating_error")) unless @order.can_backdate_order_details?
-
-          # update order detail statuses if you've changed it while acting as
-          if acting_as? && params[:order_status_id].present?
-            @order.backdate_order_details!(session_user, params[:order_status_id])
-          else
-            @order.complete_past_reservations!
-          end
-        end
+    order_purchaser.purchase!
 
-        Notifier.delay.order_receipt(user: @order.user, order: @order) if should_send_notification?
-
-        # If we're only making a single reservation, we'll redirect
-        if @order.order_details.size == 1 && @order.order_details[0].product.is_a?(Instrument) && !@order.order_details[0].bundled? && !acting_as?
-          od = @order.order_details[0]
-
-          if od.reservation.can_switch_instrument_on?
-            redirect_to order_order_detail_reservation_switch_instrument_path(@order, od, od.reservation, switch: "on", redirect_to: reservations_path)
-          else
-            redirect_to reservations_path
-          end
-
-          flash[:notice] = "Reservation completed successfully"
+    if order_purchaser.quantities_changed?
+      flash[:notice] = I18n.t("controllers.orders.purchase.quantities_changed")
+      redirect_to order_path(@order)
+    else
+      if single_reservation?
+        flash[:notice] = I18n.t("controllers.orders.purchase.reservation.success")
+        if can_switch_instrument_on?
+          redirect_to switch_instrument_path
         else
-          redirect_to receipt_order_path(@order)
+          redirect_to reservations_path
         end
-
-        return
+      else
+        redirect_to receipt_order_path(@order)
       end
-    rescue => e
-      flash[:error] = I18n.t("orders.purchase.error")
-      flash[:error] += " #{e.message}" if e.message
-      puts e.message
-      @order.reload.invalidate!
-      redirect_to(order_path(@order)) && return
     end
+  rescue NUCore::QuantityUpdateChangeException => e
```

Comment:
Why do we need both this error handling _and_ `order_purchaser.quantities_changed?`?